### PR TITLE
feat: add daily stripe subscription reconciliation job

### LINKED ISSUE #38: Add Stripe subscription reconciliation job
## Problem
If webhooks fail during an outage, the database state can diverge from Stripe's truth. There's no mechanism to detect or correct this drift.

## Solution
Add a daily scheduled Convex function that:
1. Fetches active subscriptions from Stripe API
2. Compares against database subscription states
3. Logs discrepancies (and optionally auto-corrects)

## Acceptance Criteria
- [ ] Scheduled function runs daily
- [ ] Compares Stripe subscriptions to Convex billing records
- [ ] Logs any mismatches with customer ID and state diff
- [ ] Optional: auto-update database from Stripe (source of truth)

## Priority
Medium - Improves operational resilience

## References
- `convex/billing.ts` - current subscription logic
- Stripe API: `stripe.subscriptions.list()`

### PR DESCRIPTION
**Pull request by phrazzld**
## Summary
Adds a daily Convex reconciliation job for Stripe subscription state drift detection, with optional auto-correction support through existing idempotent billing mutation paths.

Closes #38.

## Changes
- Added `internal.billing.reconcileStripeSubscriptionsInternal` action in `convex/billing.ts`.
- Added `internal.billing.listStripeBillingRecordsInternal` query for Stripe-linked billing snapshots.
- Added `convex/crons.ts` daily job at `04:15 UTC` calling reconciliation action in log-only mode (`autoCorrect: false`).
- Added pure reconciliation module `lib/billing/reconciliation.ts`:
  - Stripe status normalization (`trialing -> active`, `incomplete_expired -> expired`)
  - mismatch detection: missing in Stripe, missing in DB, status/subscription/period-end mismatch
  - proposed update payload generation
- Added unit tests in `lib/billing/__tests__/reconciliation.test.ts`.
- Updated architecture docs with reconciliation flow.

## Acceptance Criteria
- [x] Scheduled function runs daily
- [x] Compares Stripe subscriptions to Convex billing records
- [x] Logs any mismatches with customer ID and state diff
- [x] Optional: auto-update database from Stripe (source of truth)

## Manual QA
1. `bun vitest lib/billing/__tests__/reconciliation.test.ts`
   - Expected: 8 passing tests.
2. `npx convex codegen`
   - Expected: successful type generation/typecheck.
3. `bun check`
   - Expected: completes successfully (existing repo lint warnings remain unchanged).
4. `bun vitest`
   - Expected: full suite passes.
5. `bun run build`
   - Expected: production build succeeds.
6. (Optional runtime verification) run Convex function in dev with `autoCorrect: false` and verify mismatch logs in function logs.

## What Changed
```mermaid
graph TD
  A[Convex Daily Cron 04:15 UTC] --> B[reconcileStripeSubscriptionsInternal]
  B --> C[Fetch Stripe subscriptions]
  B --> D[Fetch Convex userProgress billing snapshots]
  C --> E[reconcileSubscriptionState]
  D --> E
  E --> F[Log mismatches]
  E --> G[Proposed updates]
  G --> H{autoCorrect?}
  H -- no --> I[return summary]
  H -- yes --> J[updateFromStripeInternal per customer]
  J --> I
```

## Before / After
Before: webhook-driven updates could drift from Stripe truth with no scheduled detection mechanism.
After: daily reconciliation detects drift against Stripe source-of-truth and logs actionable diffs; optional auto-correction path is available without changing webhook semantics.

## Test Coverage
- Added: `lib/billing/__tests__/reconciliation.test.ts`
  - status normalization mapping
  - exact/no-diff behavior
  - missing-in-stripe and missing-in-db detection
  - mismatch detection and proposed update generation
  - newest-subscription selection when customer has multiple Stripe subscriptions
- Existing suite run: `bun vitest` (all passing in branch).

## Summary by CodeRabbit

* **New Features**
  * Daily Stripe subscription reconciliation at 04:15 UTC: scans Stripe vs local billing, logs mismatches, and generates proposed updates; runs log-only by default.
  * Optional auto-correct flag to apply fixes; reconciliation returns summary counts (scanned, mismatches, proposed, corrected, failed).

* **Documentation**
  * Added architecture docs describing the billing reconciliation flow and behavior.

* **Tests**
  * Added comprehensive unit tests for normalization, reconciliation logic, pagination, and edge cases.

## PR Unblock Before / After
Before: Cerberus and bot reviewers had open correctness/testing findings plus unresolved scalability and resilience review threads.
After: Correctness/testing findings were fixed in commits 07774c8, 4e43250, 734d037, and ebc93c1; remaining architecture scale work is deferred to #137 and Stripe timeout/retry hardening is deferred to #139; all review threads are replied/resolved and checks are green.

## PR Fix: Self-Review Findings (dce5abd)
**Before**: `incomplete_expired` mapped to `incomplete` (actionable state) instead of `expired` (terminal state); `stale_event` mutation rejections counted as failures in reconciliation metrics.
**After**: `incomplete_expired` correctly maps to `expired`; `stale_event`/`duplicate_event` returns treated as success (data already current). Both fixes include regression tests.

## PR Fix: Pagination Cursor Bug (95f0ddb)
**Before**: If a Stripe API page contained only subscriptions with null customer IDs, all items were filtered out and `collectStripeSubscriptions` threw a pagination invariant error, aborting the entire reconciliation job.
**After**: The fetch callback now passes `lastRawId` (the last Stripe subscription ID pre-filter) as an explicit cursor, allowing pagination to continue past fully-filtered pages.

### CODE REVIEW

File: convex/billing.ts [hunk 1]
```diff
@@ -404,3 +410,119 @@ export const clearStripeCustomer: ReturnType<typeof action> = action({
...
+          typeof subscription.customer === "string"
+            ? subscription.customer
+            : subscription.customer?.id;
+        if (!customerId) continue;
+        stripeSubscriptions.push({
+          id: subscription.id,
+          customerId,
+          status: subscription.status,
+          created: subscription.created,
+          currentPeriodEnd: subscription.items.data[0]?.current_period_end,
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/billing.ts]: Classification: risk
Severity: medium
Decision: fix now.
Change: Period-end extraction now uses top-level fallback first, then item-level current_period_end.
Verification: Commit b871bf5; bun vitest and bun run build pass.

File: lib/billing/reconciliation.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    case "unpaid":
+    case "incomplete":
+      return status;
+    default:
+      return "expired";
+  }
+}
+
+function getLatestSubscriptionByCustomer(
+  subscriptions: StripeSubscriptionSnapshot[]
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply lib/billing/reconciliation.ts]: Classification: bug
Severity: high
Decision: fix now.
Change: Unknown Stripe statuses no longer default to expired; they are surfaced as unsupported_stripe_status and skipped for auto-correction.
Verification: Commit b871bf5; normalization tests updated and passing.

File: lib/billing/reconciliation.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+      return "expired";
+    case "active":
+    case "past_due":
+    case "canceled":
+    case "unpaid":
+    case "incomplete":
+      return status;
+    default:
+      return "expired";
+  }
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply lib/billing/reconciliation.ts]: Classification: bug
Severity: high
Decision: fix now.
Change: Added explicit paused handling and changed unknown-status behavior to non-destructive; aligned incomplete_expired mapping with webhook semantics.
Verification: Commit b871bf5; updated tests pass.

File: convex/billing.ts [hunk 4]
```diff
@@ -404,3 +411,125 @@ export const clearStripeCustomer: ReturnType<typeof action> = action({
...
+    const users = await ctx.db.query("userProgress").withIndex("by_stripe_customer").collect();
+    return users
+      .filter((user) => user.stripeCustomerId)
+      .map((user) => ({
+        userId: user.userId,
+        stripeCustomerId: user.stripeCustomerId as string,
+        stripeSubscriptionId: user.stripeSubscriptionId,
+        subscriptionStatus: user.subscriptionStatus,
+        currentPeriodEnd: user.currentPeriodEnd,
+      }));
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/billing.ts]: Classification: risk
Severity: medium
Decision: defer. Follow-up tracked in #136.
Change: No change in this PR; defer internalAction orchestration integration tests to focused follow-up.
Verification: Local gates are green and action logic paths are partially covered by extracted helper tests.

File: convex/billing.ts [hunk 5]
```diff
@@ -404,3 +411,125 @@ export const clearStripeCustomer: ReturnType<typeof action> = action({
     });
   },
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/billing.ts]: Classification: risk
Severity: medium
Decision: fixed
Change: Added an internal-query test for `listStripeBillingRecordsInternal` validating index usage, filtering users without `stripeCustomerId`, and output field mapping.
Verification: `bun vitest convex/__tests__/billing.test.ts` and full suite pass on commit 07774c8.
